### ISSUE TITLE
trims not working in .JEF files

### ISSUE THREAD

**Issue by lexelby** [inkstitch/inkstitch#377]
Described in https://github.com/inkstitch/inkstitch/issues/374#issuecomment-452788483, it seems that TRIM commands aren't working in JEF files.  The same design saved to DST on the same machine trims properly.

**Comment by lexelby** [comment]
Could someone create a simple design in another digitizing program and save it as DST, JEF, and PES?  It would need to have a couple of trims in it so that I can see how they're encoded in the JEF file.

**Comment by lexelby** [comment]
I just bought a 400e, so I'll be able to debug this issue directly :D

**Comment by jameskolme** [comment]
Wohoo \o/  

Not much have changed:
http://www.sewvacdirect.com/content/Comparison%20Chart%20400E%20vs%20500E.pdf 
Biggest difference is maximum hoop size, but I think 7.9” x 7.9” is plenty. I really like trim and advanced needle threader.

**Comment by lexelby** [comment]
How does the needle threader work?  Is it a lever you push down?

**Comment by jameskolme** [comment]
Yes, seems to be same threader and after I fixed it, now works every time. First it just barely worked time to time, every time small hook hit side of needle and most likely jumped out right side of needle.  There are no guides how to fix this, but hook itself is quite soft material at the bottom or place where it starts. Put small flat screwdriver exactly where hook starts and it can be bend slightly, just by twisting screwdriver slightly. I did extensive google for this and seems to be surprisingly common issue to not work properly. I was amazed there was nothing to adjust this properly, like adjustment screw or something behind plastic cover.

**Comment by jameskolme** [comment]
Trim solution:
https://github.com/EmbroidePy/pyembroidery/issues/54

**Comment by lexelby** [comment]
Oh wow, this machine is so nice :D 

I just got it and gave it a go.  Super fast, and like you said, trims work great in DST files.  I agree that the thing where it doesn't read trims from the file in the auto setting is weird, but whatever.  It looks like if you do manual setup and set the jump trim to the max, it just turns off jump trimming, and then you can have only cut commands.  Works for me.  I'll get the pyembroidery fix merged tomorrow.

About the thread trimmer: I think I need some practice.  It seems like most of the time it just tears the thread.  Any tips?  I'm considering just using a separate wire needle threader :P

**Comment by jameskolme** [comment]
I have noticed at certain thread, it will tear it sometimes, hook wants to hold maybe too much, but I'm not complaining :) It might help if thread is not fully inserted to last ymmmm "holding place or gripper"  Last gap before routing thread to manual cutting blade. 

DST works directly, only thing is, I can make bad DST files too without actual error messages. Not sure if Janome is super picky about file formatting. Something to do with auto route (possibly), but let's fix JEF first and see what is what. I have excellent QR-code ready for test.

**Comment by lexelby** [comment]
I was doing it wrong. :)  As you know, one is supposed to pass the thread from left to right under the hook, rather than right to left as I had been doing.  Works great now.

Sorry I still haven't gotten to the bottom of this JEF issue.  Things have been super busy here, but I'm hoping to spend some time working on this soon.

**Comment by lexelby** [comment]
Interesting.  Looking at the JEF file from Embroideryware in a hex editor, it looks to me like it's encoded as 3 colors, but all with the same thread color.  In other words, it's encoding a TRIM as if it's a COLOR CHANGE to the same color.  I do see the 0-length jump that tatarize was talking about, but that may be a red herring.

@jameskolme Could you create a JEF file that is 3 lines of different colors?  I want to see what a normal color change looks like.

**Comment by jameskolme** [comment]
[3 colors.zip](https://github.com/inkstitch/inkstitch/files/2828451/3.colors.zip)

I'm wondering is there anything else affected, trim using color change or do actual programmed trim.

Tatarize made it work, hopefully it was this:
[3chv-trim.zip](https://github.com/inkstitch/inkstitch/files/2828518/3chv-trim.zip)

**Comment by jameskolme** [comment]
Directly from support: just alternate between "TopStitches" and "TackDown" and thread trim will be placed.

(If lucky!)

Sometimes dummy stitches are needed to trigger trim when needed and sometimes there is no way to trim anything. Before this gets 100 pages long rant, I just like to refer this program as my embroidery learning money.

**Comment by lexelby** [comment]
Okay, I got it!  Just like in a DST file, three jumps in a row equals a trim in a JEF.  In fact, it actually seems to be three "commands" in a row that trigger trimming.  So "color change, jump, jump" will cause a trim.  

This holds true even if the color change is to the same color.  If the JEF file has a  color change from black to black, the machine will not count that as an actual color change; it just continues sewing.

Just doing a color change to the same color is not enough though.  There have to be at least two trims afterward.  Embroideryware throws in five jumps after the pretend color change (just to be sure, I guess?).

Those jumps are allowed to be zero-length.  That's why tatarize originally thought a zero-length jump is a color change, because embroideryware does use zero-length jumps to ensure that it has enough jumps in a row.  This also explains why the current code only _sometimes_ trims: it will only trim if there are a couple more jumps after the trim point, which would force three in a row.

There's one more detail that I've discovered: it's not possible to have two stitches more than 2 inches apart (exactly 2.0") without triggering a trim.  The only way to do it is to turn off "trim at cutting command" in the settings.  This is because jumps and stitches of 12.7mm, and 2.0 inches is the furthest you can separate stitches before you're forced to put three jumps in a row to get from one stitch to the next.

Okay, I got a little into the weeds there, sorry!

Bottom line is, I've solved it, and now I just need to code it.  There really is no substitute for having a real machine in front of me to experiment with!

**Comment by jameskolme** [comment]
Excellent work. Have you tested does pause command work?

**Comment by lexelby** [comment]
As far as I know, single-needle machines don't support a pause command.  You can just create a color change in your design which will pause the machine.

Of course, Janome does have multi-needle machines like the MB-4S, and I believe they also take JEF files.   From reading the instruction manual awhile back I think I remember that you assign a pause at the machine, by setting a given color change as a pause. So Ink/Stitch's normal pause handing will work fine, because it inserts an extra color change.

Well, there is one issue.  I've seen that the 400e ignores a color change of you change to the same color.  JEF has a specific palette of thread colors that it supports.  If you have two consecutive colors in your design that are very similar, so that they get encoded in the file as the same color, I bet that the  machine will ignore the color change.  I'll need to make pyembroidery detect that and shift one of the colors just a bit to force a real color change.

Does anyone have access to an MB-4S or MB-7 that we can test with?

**Comment by jameskolme** [comment]
This is not an issue or anything, but I noticed strange behavior.  Machine Janome 500e, if programmed trims are present on program and there still are places where auto trim should work, no programmed trims and jumps over 3mm. This situation makes auto trimming very unreliable, might work on some places and simply jump others.  So quick copy/paste with known auto trim program is not good idea to mix it in with programmed trim program. This is probably why programmed trims are disabled by default, undocumented feature or something. Attach trim command now works nicely so this is not so big issue and is related to machine anyway.

**Comment by lexelby** [comment]
Huh, that's weird.  Is that the case only for files produced with Ink/Stitch, or with files from other software too?  I want to make sure this isn't a case of Ink/Stitch still doing something wrong in the JEF file format.
